### PR TITLE
Rename base libraries var

### PR DESCRIPTION
**Pull request by aurianer**
- Transform the current `HPX_LIBRARIES` cache variable which was linking only base libraries like `rt`, `dl`, etc. to the `hpx_base_libraries` target leaving the basic `HPX_LIBRARIES` to be used in other projects
- Remove duplicated link to external asio in `HPX_SetupBoost`
- Fix typo that resulted in a bug in `hpx_add_example_target_dependencies`
- Fix some examples cmake dependencies
- Remove the use of `hpx_library_dir` and instead `target_link_directories`
- Refactor CUDA setup

### CODE REVIEW

**Comment by hkaiser** [comment]
Wouldn't this break dependent projects that were relying on `HPX_LIBRARIES`?

**Comment by msimberg** [comment]
> Wouldn't this break dependent projects that were relying on `HPX_LIBRARIES`?

Depends what "relying on" means. I always imagined the use case was `(target_)link_libraries(... ${HPX_LIBRARIES})` and that still works perfectly with this. Same for `HPX_INCLUDE_DIRS` etc. (since they will be propagated from the targets in `HPX_LIRBARIES`). If someone was relying on `HPX_LIBRARIES` containing full paths and library names (e.g. for pkgconfig file generation) it won't work. Are you doing that in phylanx? Have we made any promises about what `HPX_LIBRARIES` can be used for?

**Comment by hkaiser** [comment]
@msimberg all is well, I just wanted to make sure this wouldn't break anything down the road. Thanks!

**Comment by msimberg** [comment]
> @msimberg all is well, I just wanted to make sure this wouldn't break anything down the road. Thanks!

Your suspicion is warranted ;) Everything *should* be okay, but I may of course have missed something. We will find out...

**Comment by aurianer** [comment]
Just changed the `HPX_BASE_LIBRARIES` into the target `hpx_base_libraries`

**Comment by msimberg** [comment]
@hkaiser by the way, I missed initially the context for your question. I thought your comment was on #4176. But I'm sure you got by now that renaming or removing `HPX_LIBRARIES` is ok because of #4176...

**Comment by msimberg** [comment]
@aurianer APEX still relies on the `hpx_libraries` function: https://circleci.com/gh/STEllAR-GROUP/hpx/140775?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-checks-link. Since APEX is essentially an HPX module now we can probably go and do the necessary changes to APEX itself.

**Comment by aurianer** [comment]
Actually it doesn't on `origin/develop` but I'm not sure which tag the tests are fetching

File: cmake/HPX_SetupCUDA.cmake [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+#
+# SPDX-License-Identifier: BSL-1.0
+# Distributed under the Boost Software License, Version 1.0. (See accompanying
+# file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+
+if(HPX_WITH_CUDA)
+
+  find_package(CUDA REQUIRED)
+  set(HPX_WITH_COMPUTE ON)
```
Comment (msimberg) [inline cmake/HPX_SetupCUDA.cmake]: I wonder if this one has to be `PARENT_SCOPE`... let's give it a try and fix it afterwards if needed.

Comment (aurianer) [reply cmake/HPX_SetupCUDA.cmake]: When calling `include(<cmake script>)` the scope stays identical and as `HPX_SetupCuda.cmake` is called in the top `CMakeLists.txt`, I would say there is no `PARENT_SCOPE` ^^

Comment (msimberg) [reply cmake/HPX_SetupCUDA.cmake]: Indeed, very good point. Thanks!

**Pull request merged**